telco-features: document a PTP BC on GNR-D (#1130)
suse-edge/suse-edge.github.io: asciidoc/product/atip-features.adoc

====

At this point, your hosts should have a working and running PTP stack and will start negotiating their PTP role.

==== PTP on Intel Granite Rapids-D platforms

> SUSE Edge supports precise timing on a variety of different platforms, including the latest Intel Granite Rapids-D based designs (GNR-D). These new servers typically include an embedded high-speed Ethernet controller and one or more add-in cards with an Ethernet controller for additional ports. Contrary to previous platform generations, they are also equipped with a dedicated DPLL allowing the routing of the timing signals across the system and the different NICS, no longer requiring external cabling.

Reviewer: ```suggestion
SUSE Edge supports precise timing on a variety of different platforms, including the latest Intel Granite Rapids-D based designs (GNR-D). These new servers typically include an embedded high-speed Ethernet controller and one or more add-in cards with an Ethernet controller for additional ports. Contrary to previous platform generations, they are also equipped with a dedicated DPLL that allows the routing of the timing signals across the system and the different NICS, and no longer requires external cabling.
```

The overall deployment is thus simplified, but specific steps and software components are required, such as boot time DPLL and internal signal configuration.

Before going through the setup steps, download and install the latest kernel driver for the _Intel 800 Series Network Devices_ from the https://www.intel.com/content/www/us/en/download/19630/intel-network-adapter-driver-for-800-series-devices-under-linux.html[Intel website]. Select the zip file containing the "ice_RPM_Files" and unpack the archive, then install the `ice-kmp-default-*sles16sp0.x86_64.rpm` RPM there contained, for example file `ice-kmp-default-2.4.5_k6.12.0_160000.5-1.sles16sp0.x86_64.rpm`:

[,console]
----
# transactional-update pkg install ice-kmp-default-2.4.5_k6.12.0_160000.5-1.sles16sp0.x86_64.rpm
----

Reboot the server to make the change persistent and effective.

[NOTE]
====
* A Dell PowerEdge XR8720t system will be used as the reference system through the examples, the kernel interfaces and port configuration might be different on a different GNR-D server.
* A Telco deployment is assumed and the Telco profile G.8275.1 used through the following sections for PTP.
====

===== PTP Boundary Clock

> This section describes the process of setting up a Boundary Clock (BC) on a GNR-D system, where one port on the integrated NIC receives the PTP timestamps from a GM and all the remaining ports in the system act as Time Transmitters towards other network nodes. Please note that the setup depicted here is completely static, if the GM becomes unavailable the downstream synchronization will exclusively rely on the host holdhover capabilities; no automatic switching to a different source of time reference will be performed.

Reviewer: ```suggestion
This section describes the process of setting up a Boundary Clock (BC) on a GNR-D system, where one port on the integrated NIC receives the PTP timestamps from a GM and all the remaining ports in the system act as Time Transmitters towards other network nodes.
[NOTE]
====
The setup described here is completely static. If the GM becomes unavailable, the downstream synchronization will exclusively rely on the host holdhover capabilities; no automatic switching to a different source of time reference will be performed.
====
```

> To enable the BC behavior, execute the following steps in the same order every time the system is booted:

Reviewer: ```suggestion
To enable the BC behavior, run the following steps in the same order every time the system is booted:
```

> . Verify the expected PTP devices for controlling the PHCs are present:

Reviewer: ```suggestion
. Verify the expected PTP devices that control the PHCs are present:
```

+
[,console]
----
# ls -l /sys/class/ptp/
total 0
lrwxrwxrwx. 1 root root 0 Apr 27 20:21 ptp1 -> ../../devices/pci0000:12/0000:12:04.0/0000:13:00.0/ptp/ptp1
lrwxrwxrwx. 1 root root 0 Apr 27 20:21 ptp2 -> ../../devices/pci0000:6b/0000:6b:02.0/0000:6c:00.0/ptp/ptp2
lrwxrwxrwx. 1 root root 0 Apr 27 20:21 ptp3 -> ../../devices/pci0000:6b/0000:6b:06.0/0000:6e:00.0/ptp/ptp3
----
+
[NOTE]
====
Adjust the subsequent commands according to the actual names found on the system being configured.
====

. Identify the embedded controller, as it will require specific configuration:
+

> The previous output should provide an indication of the device they are bound to through the PCI address and `lspci` can be used to look up the device type:

Reviewer: ```suggestion
From the previous output, 
The previous output should indicate the device to which they are bound through the PCI address and  you can use `lspci` to look up the device type:
```

Author: Correct me if I'm wrong but I guess the "From the previous output," is a leftover from previous editing, isn't it?

Reviewer: Sorry. I must have missed removing that line! Looks like you have taken care of that.

+
[,console]
----
# lspci -s 0000:13:00.0
13:00.0 Ethernet controller: Intel Corporation Ethernet Connection E825-C for SFP (rev 04)
# lspci -s 0000:6c:00.0
6c:00.0 Ethernet controller: Intel Corporation Ethernet Controller E830-CC for SFP
# lspci -s 0000:6e:00.0
6e:00.0 Ethernet controller: Intel Corporation Ethernet Controller E830-CC for SFP
----
+
This specific server is equipped with an _Intel E825-C_ embedded controller and two _E830-CC_ add-in cards.
+

> Alternatively it is possible to find out more information about the controller by looking up the first network port of the controller that exposes that PHC:

Reviewer: ```suggestion
Alternatively, you can find more information about the controller by looking up the first network port of the controller that exposes that PHC:

```

+
[,console]
----
# ls -l /sys/class/ptp/ptp0/device/net/
total 0
drwxr-xr-x. 5 root root 0 Apr 30 15:16 em5
# ethtool -i em5
driver: ice
version: 2.4.5
firmware-version: 4.03 0x80007f91 1.3881.0
expansion-rom-version:
bus-info: 0000:13:00.0
[...]
----
+

> Note that `em5` is the first port on controlled by the integrated controller on this system.

Reviewer: What does "on controlled" mean here?

Author: Good catch, some incomplete rewording.

. Configure the signals between the PHC on the integrated Ethernet controller and the system timing DPLL:

> .. Let the DPLL drive the Ethernet clock on the integrated controller

Reviewer: ```suggestion
.. Let the DPLL drive the Ethernet clock on the integrated controller:
```

+
[,console]
----
# echo 4 1 > /sys/class/net/em5/device/tspll_cfg
----

.. Enable a pulse every top of second (Pulse Per Second, PPS) from the PHC to the DPLL through the SDP2 pin (1kHz signal):
+
[,console]
----
# echo 2 1 > /sys/class/ptp/ptp1/pins/SDP2
# echo 1 0 0 0 1000000 > /sys/class/ptp/ptp1/period
----

.. Optionally, pin SPD0 (1Hz signal) can be enabled too:
+
[,console]
----
# echo 2 2 > /sys/class/ptp/ptp1/pins/SDP0
# echo 2 0 0 1 0 > /sys/class/ptp/ptp1/period
----

.. Enable the 1 PPS signal to be propagated from the timing system DPLL to the additional add-in cards:
+
[,console]
----
# echo 1 1 > /sys/class/ptp/ptp2/pins/SDP1
# echo 1 1 > /sys/class/ptp/ptp3/pins/SDP1
----

. Run `ts2phc` to keep PHCs of the add-in cards in sync with the same timestamp of PHC `/dev/ptp1`:
+
[,console]
----
# ts2phc -f ts2phc-aic-all.cfg -s /dev/ptp1
----
+

> Refer to <<gnrd-ts2phc-config-aics>> for the content of `ts2phc-aic-all.cfg`.

Reviewer: For consistency, either you mention "refer" or "check". 
For this section, "refer" is more appropriate.

. Run an instance of `ptp4l` for the integrated controller (`/dev/ptp1`):
+
[,console]
----
# ptp4l -f ptp4l-G.8275.1-nac.conf
----
+
Check section <<gnrd-ptp4l-config-nac>> for the content of the configuration file.

. Run an instance of `ptp4l` for each add-in card (`/dev/ptp3` and/or `/dev/ptp2`):
+
[,console]
----
# ptp4l -f ptp4l-G.8275.1-aic1.conf
# ptp4l -f ptp4l-G.8275.1-aic2.conf
----
+
Check section <<gnrd-ptp4l-config-aics>> for the content of the configuration file.

> Note that each of these instances must use different control and read-only sockets not to conflict with each other and enable the next step.

Reviewer: ```suggestion
[NOTE]
====
To prevent conflicts and enable the next step, each of these instances must use different control and read-only sockets.
====
```

> . Synchronize the PTP runtime parameters and values of the AICs ptp4l instances with `pmc`:

Reviewer: ```suggestion
. Synchronize the PTP runtime parameters and values of the AICs `ptp4l` instances with `pmc`:
```

+
A BC is expected to forward not only the time of a GM it is connected to, but also ancillary parameters, such as its clock class and accuracy level. This happens automatically within a group of interfaces managed by the same `ptp4l` instance, but not accross different `ptp4l` instances, handling different PHCs and ports.
+

> So, once the system is synchronized to a GM, query the `ptp4l` instance on the integrated controller and set the values on the remaining ptp4l instances via `pmc` commands. For example:

Reviewer: ```suggestion
Once the system is synchronized to a GM, query the `ptp4l` instance on the integrated controller and set the values on the remaining `ptp4l` instances via `pmc` commands. For example:
```

                currentUtcOffsetValid 0
                ptpTimescale          1
                timeTraceable         0
                frequencyTraceable    0
                timeSource            0xa0
pmc -f /etc/ptp4l-G.8275.1-aic1-static.conf -u -b 0 'SET GRANDMASTER_SETTINGS_NP
                clockClass              8
                clockAccuracy           0xfe
                offsetScaledLogVariance 0xffff
                currentUtcOffset        37
                leap61                  0
                leap59                  0
                currentUtcOffsetValid   0
                ptpTimescale            1
                timeTraceable           0
                frequencyTraceable      0
                timeSource              0xa0
----
+
Check <<gnrd-pmc-forward>> to automate this process and avoid manual steps.

. Optionally run `phc2sys` to derive the system time from the PTP synchronization:
+
[,console]
----
# phc2sys -f /etc/ptp4l-G.8275.1-nac.conf -s /dev/ptp1 -c CLOCK_REALTIME -w
----
+
The input file is the same configuration file used for `ptp4l` and the integrated controller. See <<ptp-phc2sys-setup>> for letting systemd handle its execution.

> ===== Configuration files

Reviewer: Is there any introduction or summary for this section?

Author: It was meant to be an Appendix like section, but we don't have this structure yet. I'll add one.

[#gnrd-ptp4l-config-nac]
====== `ptp4l` configuration file for the integrated controller (`/dev/ptp1`)

[source,]
----
[global]
domainNumber                    24
priority2                       255
dataset_comparison              G.8275.x
G.8275.portDS.localPriority     128
G.8275.defaultDS.localPriority  128
maxStepsRemoved                 255
logAnnounceInterval             -3
logSyncInterval                 -4
logMinDelayReqInterval          -4
announceReceiptTimeout          3
ptp_dst_mac                     01:1b:19:00:00:00
network_transport               L2
uds_address                     /var/run/ptp4l-nac
uds_ro_address                  /var/run/ptp4lro-nac

[em1]

[em2]

[em3]

[em4]
----

> The above configuration leverages the BMCA to determine the role of the port, assuming a GM is available through one of them. If tighter control is needed, add the `serverOnly ` flag under each interface definitione. For more details, refer to <<ptp-bc-setup>> or to the add-in cards configuration below as an example.

Reviewer: ```suggestion
The above configuration leverages the BMCA to determine the role of the port, assuming a GM is available through one of them. If tighter control is needed, add the `serverOnly` flag under each interface definition. For more details, refer to <<ptp-bc-setup>> or to the add-in cards configuration below as an example.
```

[#gnrd-ptp4l-config-aics]
====== `ptp4l` configuration file for add-in cards (`/dev/ptp2` or `/dev/ptp3`)

[source,]
----
# Telecom G.8275.1 example configuration
[global]
domainNumber                    24
priority2                       255
dataset_comparison              G.8275.x
G.8275.portDS.localPriority     250
G.8275.defaultDS.localPriority  250
maxStepsRemoved                 255
logAnnounceInterval             -3
logSyncInterval                 -4
logMinDelayReqInterval          -4
announceReceiptTimeout          3
ptp_dst_mac                     01:1b:19:00:00:00
network_transport               L2
uds_address                     /var/run/ptp4l-aic1
uds_ro_address                  /var/run/ptp4lro-aic1

[p1p1]
serverOnly                      1

[p1p2]
serverOnly                      1

[p1p3]
serverOnly                      1

[p1p4]
serverOnly                      1
----

> For the sake of brevity only four interfaces are listed, adjust according to your needs. All the interfaces have been statically set to the TimeTransmitter role.

Reviewer: ```suggestion
For the sake of brevity, only four interfaces are listed. You need to adjust according to your requirements. All the interfaces have been statically set to the TimeTransmitter role.
```

[#gnrd-ts2phc-config-aics]
====== `ts2phc` configuration file for add-in cards

[source,]
----
[global]
use_syslog              0
verbose                 1
logging_level           7
ts2phc.pulsewidth       100000000

[p1p1]
ts2phc.channel          1
ts2phc.extts_polarity   rising
ts2phc.pin_index        1

[p2p1]
ts2phc.channel          1
ts2phc.extts_polarity   rising
ts2phc.pin_index        1
----

This configuration covers the synchronization of two add-in cards (referenced by their first port, `p1p1` and `p2p1`) via the SDP1 on each controller.

[#gnrd-pmc-forward]
====== GM parameters forwarding script

The script below can be leveraged to forward the GM values and parameters to Time Receivers on ports handled by `ptp4l` instances other than the one where the GM lives. Do not forget to edit the `Input variables` section to adjust the file names and run it once the system is sychronized to a GM.

[bash,]
----
#!/bin/bash

> #

Author: Do we need to add a copyright here?

Reviewer 2: Perhaps @ranjinimn can comment on that - I would say no, because the docs as a whole are published with a copyright note in the footer e.g see https://documentation.suse.com/suse-edge/3.5/html/edge/index.html

Reviewer: Agree with @hardys

Author: I wonder however if it would be better to, as this will likely end up in the examples as a standalone file.

+
[,console]
----
# echo 2 1 > /sys/class/ptp/ptp1/pins/SDP2
# echo 1 0 0 0 1000000 > /sys/class/ptp/ptp1/period
----

.. Optionally, pin SPD0 (1Hz signal) can be enabled too:
+
[,console]
----
# echo 2 2 > /sys/class/ptp/ptp1/pins/SDP0
# echo 2 0 0 1 0 > /sys/class/ptp/ptp1/period
----

.. Enable the 1 PPS signal to be propagated from the timing system DPLL to the additional add-in cards:
+
[,console]
----
# echo 1 1 > /sys/class/ptp/ptp2/pins/SDP1
# echo 1 1 > /sys/class/ptp/ptp3/pins/SDP1
----

. Run `ts2phc` to keep PHCs of the add-in cards in sync with the same timestamp of PHC `/dev/ptp1`:
+
[,console]
----
# ts2phc -f ts2phc-aic-all.cfg -s /dev/ptp1
----
+
Refer to <<gnrd-ts2phc-config-aics>> for the content of `ts2phc-aic-all.cfg`.

. Run an instance of `ptp4l` for the integrated controller (`/dev/ptp1`):
+
[,console]
----
# ptp4l -f ptp4l-G.8275.1-nac.conf
----
+
Refer to <<gnrd-ptp4l-config-nac>> for the content of `ptp4l-G.8275.1-nac.conf`.

. Run an instance of `ptp4l` for each add-in card (`/dev/ptp3` and/or `/dev/ptp2`):
+
[,console]
----
# ptp4l -f ptp4l-G.8275.1-aic1.conf
# ptp4l -f ptp4l-G.8275.1-aic2.conf
----

> +
> Refer to <<gnrd-ptp4l-config-aics>> for the content of the configuration files.
> [NOTE]
> ====
> To prevent conflicts and enable the next step, each of these instances must use different control and read-only sockets.
> ====

Reviewer: ```suggestion
Refer to <<gnrd-ptp4l-config-aics>> for the content of the configuration files.
+
[NOTE]
====
To prevent conflicts and enable the next step, each of these instances must use different control and read-only sockets.
====
```

Author: Ah, thank you! 👍

At this point, your hosts should have a working and running PTP stack and will start negotiating their PTP role.

=== PTP on Intel Granite Rapids-D platforms

> SUSE Edge supports precise timing on a variety of different platforms, including the latest Intel Granite Rapids-D based designs (GNR-D). These new servers typically include an embedded high-speed Ethernet controller and one or more add-in cards with an Ethernet controller for additional ports. Contrary to previous platform generations, they are also equipped with a dedicated DPLL that allows the routing of the timing signals across the system and the different NICS, and no longer requires external cabling.

Reviewer 2: Sorry just noticed this product name, since #1097 merged this should be `{product}` so this renders with SUSE Telco Cloud in the product docs (which is the only version containing this content, but lets be consistent with the product variable instead of hard-coding SUSE Telco Cloud) - see the rendered preview at https://suse-edge.github.io/

Author: Absolutely right, I actually a note I forgot about to double-check if it was correct. Thanks.
